onnxruntime/core/providers/cpu/tensor/identity_op.h

```diff
@@ -23,10 +24,31 @@ class IdentityOp final : public OpKernel {
   }
 
   Status Compute(OpKernelContext* context) const override {
-    auto X_ml_type = context->InputType(0);
-    if (X_ml_type == DataTypeImpl::GetType<Tensor>()) {
-      const auto* X = context->Input<Tensor>(0);
-      ORT_ENFORCE(X != nullptr);
+    const auto* input_type_proto = Node().InputDefs()[0]->TypeAsProto();
+
+    const auto* input_ort_value = context->GetInputOrtValue(0);
+
+    // Only Optional type can be None (i.e.) not have data
+    if (input_type_proto->has_optional_type() && !input_ort_value->IsAllocated()) {
+      // We can't rely on the input OrtValue containing type information
+      // as it could be a main graph input which will be missing the type
+      // in the corresponding OrtValue for the "None" case because
+      // the user doesn't provide any input for the "None" case.
+
```

Comment:
Looks like this logic is repeated in loop.cc. Can we use a helper to avoid the duplication?

Reply:
Used a common helper to share the code in both places